src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -113,84 +114,82 @@ public void Start<TContext>(IHttpApplication<TContext> application)
+                    {
+                        var parsedAddress = ServerAddress.FromUrl(address);
+
+                        if (parsedAddress.IsUnixPipe)
                         {
-                            _disposables.Push(engine.CreateServer(ipv4Address));
+                            listenOptions.Add(new ListenOptions(parsedAddress.UnixPipePath)
+                            {
+                                Scheme = parsedAddress.Scheme,
+                                PathBase = parsedAddress.PathBase
+                            });
                         }
-                        catch (AggregateException ex) when (ex.InnerException is UvException)
+                        else
                         {
-                            var uvEx = (UvException)ex.InnerException;
-                            if (uvEx.StatusCode == Constants.EADDRINUSE)
+                            if (string.Equals(parsedAddress.Host, "localhost", StringComparison.OrdinalIgnoreCase))
                             {
-                                throw new IOException($"Failed to bind to address {parsedAddress.ToString()} on the IPv4 loopback interface: port already in use.", ex);
+                                // "localhost" for both IPv4 and IPv6 can't be represented as an IPEndPoint.
+                                StartLocalhost(engine, parsedAddress);
+
+                                // If StartLocalhost doesn't throw, there is at least one listener.
+                                // The port cannot change for "localhost".
+                                _serverAddresses.Addresses.Add(parsedAddress.ToString());
+                                atLeastOneListener = true;
                             }
                             else
                             {
-                                _logger.LogWarning(0, $"Unable to bind to {parsedAddress.ToString()} on the IPv4 loopback interface: ({uvEx.Message})");
-                                exceptions.Add(uvEx);
+                                // These endPoints will be added later to _serverAddresses.Addresses
+                                listenOptions.Add(new ListenOptions(CreateIPEndPoint(parsedAddress))
+                                {
+                                    Scheme = parsedAddress.Scheme,
+                                    PathBase = parsedAddress.PathBase
+                                });
                             }
                         }
+                    }
+                }
 
-                        var ipv6Address = parsedAddress.WithHost("[::1]");
+                foreach (var endPoint in listenOptions)
+                {
+                    atLeastOneListener = true;
 
-                        try
+                    try
+                    {
+                        _disposables.Push(engine.CreateServer(endPoint));
+                    }
+                    catch (AggregateException ex)
```

Comment:
Not sure what you mean. This is the [same logic](https://github.com/aspnet/KestrelHttpServer/pull/1280/files#diff-2a47c9b99ecab7d9ea77789228f7e771L127) we had before for when anything other than "localhost" was specified.

We don't swallow any exceptions like we did when "localhost" was specified since there is no desire to gracefully handle binding only to the IPv4 loopback adapter and not the IPv6 one or vice-versa.